winit/src/platform_specific/wayland/subsurface_widget.rs

```diff
@@ -464,11 +500,65 @@ impl SubsurfaceState {
             subsurface.unmap();
             self.unmapped_subsurfaces.push(subsurface);
         }
+        let needs_sorting = subsurfaces.len() < view_subsurfaces.len()
+            || !self.new_iced_subsurfaces.is_empty();
+
         // Create new subsurfaces if there aren't enough.
         while subsurfaces.len() < view_subsurfaces.len() {
             subsurfaces.push(self.create_subsurface(parent));
         }
-        // Attach buffers to subsurfaces, set viewports, and commit.
+        if needs_sorting {
+            let mut sorted_subsurfaces: Vec<_> = view_subsurfaces
+                .iter()
+                .zip(subsurfaces.iter_mut())
+                .map(|(_, instance)| {
+                    (
+                        instance.parent.clone(),
+                        instance.wl_subsurface.clone(),
+                        instance.wl_surface.clone(),
+                        instance.z,
+                    )
+                })
+                .chain(self.new_iced_subsurfaces.clone().into_iter().map(
+                    |(_, parent, _, wl_subsurface, wl_surface, z)| {
+                        (parent.clone(), wl_subsurface, wl_surface, z)
+                    },
+                ))
+                .chain(ICED_SUBSURFACES.with(|surfaces| {
+                    let b = surfaces.borrow();
+                    let v: Vec<_> = b
+                        .iter()
+                        .map(move |s| {
+                            (s.1.clone(), s.3.clone(), s.4.clone(), s.5)
+                        })
+                        .collect();
+                    v.into_iter()
+                }))
+                .collect();
+
+            sorted_subsurfaces.sort_by(|a, b| {
+                let a_id = a.0.protocol_id();
+                let b_id = b.0.protocol_id();
+                (a.3, a_id).cmp(&(b.3, b_id))
+            });
```

Comment:
This would be equivalent to `sorted_subsurfaces.sort_by_key(|s| (s.3, s.0.protocol_id()))`, I think?

I'm not sure why `protocol_id()` is needed here. The order will be consistent with or without it since it isn't an unstable sort, right? But object ID re-use would make the protocol ID unreliable for anything other than defining a consistent but non-meaningful order.

Reply:
Ya, that was the idea, but I guess it is unnecessary with the stable sort.